Add/filter params to queries

psealock opened the pull request:
Create a function `getFilterQuery` to gather results from filters, comparisons, and advancedFilters to apply correct query parameters on data requests.

## Testing

### Simple Filters

These will update the tables correctly, but I think it would be nice to update the tab on the chart (in a separate issue).

<img width="336" alt="screen shot 2018-10-24 at 10 42 23 pm" src="https://user-images.githubusercontent.com/1922453/47422259-b558b000-d7de-11e8-9404-6496080a98b1.png">

1. `/wp-admin/admin.php?page=wc-admin#/analytics/products?filter=top_sales`.
2. The query should include `orderBy: 'gross_revenue'`,  `order: 'desc'`.
3. Repeat for "Top Products By Items Sold".

### Compare Filters

<img width="465" alt="screen shot 2018-10-24 at 10 50 06 pm" src="https://user-images.githubusercontent.com/1922453/47422471-3152f800-d7df-11e8-94b3-ed0f25ee72ea.png">

1.` /wp-admin/admin.php?page=wc-admin#/analytics/products?filter=compare-product`.
2. Choose at least two products.
3. Click "Compare".
4. See `products: 64,66` included in the requests.
5. The table and chart data will reflect the filter.
6. Repeat for comparing categories.

### Single Filters

<img width="422" alt="screen shot 2018-10-24 at 10 52 06 pm" src="https://user-images.githubusercontent.com/1922453/47422588-76772a00-d7df-11e8-8a0d-6a7c21cdfdca.png">

1.` /wp-admin/admin.php?page=wc-admin#/analytics/products?filter=compare-product`.
2. Choose a product.
4. See `products: 66` included in the requests and see the Table and Chart results filtered accordingly.

### Advanced Filters

<img width="815" alt="screen shot 2018-10-24 at 10 54 06 pm" src="https://user-images.githubusercontent.com/1922453/47422708-c6ee8780-d7df-11e8-8e0f-9b2e52f167e2.png">

_**Note:** Many of the Advanced Filters parameters have not yet been implemented on the back end and will not actually filter results_. Testing is best achieved through the Network tab and making sure requests to `/reports/orders` and `/reports/orders/stats` have the correct parameters added.

<img width="256" alt="screen shot 2018-10-24 at 10 48 16 pm" src="https://user-images.githubusercontent.com/1922453/47422362-f05ae380-d7de-11e8-8f38-aef4f22b73d2.png">

1. `/wp-admin/admin.php?page=wc-admin#/analytics/orders?filter=advanced`.
2. Select filters and add items to filter.
3. See the parameters included in requests. These will be the same ones found in the url.
4. Make sure the `match` parameter is passed as well.

### Additional things to consider

* Query parameters having nothing to do with filters should not be included in requests
* Back end APIs aren't finished, so the exact terms may change
* If you're reading this far, thank you!

psealock committed:
Filters: Add params to request query

belcherj commented on client/components/filters/filter/index.js:
```diff
@@ -16,6 +16,7 @@ import AnimationSlider from 'components/animation-slider';
 import DropdownButton from 'components/dropdown-button';
 import Search from 'components/search';
 import { updateQueryString } from 'lib/nav-utils';
+import { flatenFilters } from './utils';
```
Nice!

belcherj reviewed (approved):
Tested and works as expected.  Still lots to do on the front end to make all this work.  The charts and summaries are not displaying correctly for the filtered data.

LevinMedia commented:
@psealock one small fix - can you update the last filter to read "Top Products by Gross Revenue" 

> These (top items sold / gross revenue )  will update the tables correctly, but I think it would be nice to update the tab on the chart (in a separate issue).

That would be a nice touch. There's some additional behavior for how this will manifest itself in the chart and tabular data as well - The short of it is, clicking on either of the top products filters will automatically create a  comparison of the top products (ordered by items sold, or gross revenue) The chart should change to a comparison chart that has one extra feature, a setting to let you control how many products to include in the comparison. 

<img width="1070" alt="screen shot 2018-10-24 at 11 48 19 am" src="https://user-images.githubusercontent.com/4500952/47453981-dab7d080-d782-11e8-88de-24ab6322ab05.png">

The tabular data will only display the selected number of top products, and will be ordered by either items sold, or gross revenue, depending on the selected filter.  Lastly, the summary numbers will reflect just the data from the top products.

LevinMedia commented:
> Compare Filters

Looking good! Just making note that the chart should also be switched out from the standard chart to the comparison chart when it's ready.

Aljullu reviewed (commented):
This looks really great. Good job @psealock! :clap: 

I added some small comments, but it's looking good!

ryelle commented on client/analytics/report/orders/config.js:
```diff
@@ -18,7 +18,7 @@ export const filters = [
 
 /*eslint-disable max-len*/
-export const advancedFilterConfig = {
+export const advancedFilters = {
```
I would suggest leaving the individual `const`s exported (`export const advancedFilters = …`), and when you want both in one object, you can do 

```
import * as ordersConfig from 'analytics/report/orders/config';
import * as productsConfig from 'analytics/report/products/config';
```

psealock commented on client/analytics/report/products/index.js:
```diff
@@ -55,6 +55,7 @@ export default compose(
 
 		const { getProducts, isGetProductsError, isGetProductsRequesting } = select( 'wc-admin' );
+		const filterQuery = getFilterQuery( 'products', query );
 		const tableQuery = {
```
@Aljullu what do you think about using `withSelect` in `ProductsReportTable` and moving all the relevant parts such as `isProductsError` etc? We wouldn't have to pass both query and tableQuery and there would be more clarity around `isProductsError` only referring to the request for table data, eg `/reports/products` not `/reports/products/stats`.

Aljullu replied:
> what do you think about using `withSelect` in `ProductsReportTable`

I created an issue for that and assigned it to myself: #703.

psealock commented:
Thank you @Aljullu @ryelle for the comments. I've addressed them including the following fixes:

* Removed default export from configs.
* Added `tableQuery` prop to `ProductsReportTable`.

Aljullu reviewed (approved):
Everything looks good to me! :rocket:
